### PR TITLE
Add Karafka monitoring

### PR DESCRIPTION
**Pull request by dannyd4315**
### What?

This PR adds monitoring for Karafka 2.0. There are a lot of changes in how things are instrumented in the Karafka update. So we are not able to track some of the older types of metrics. 

We have run through the metrics with @robertomiranda and agreed on the metric changes.

### Why?

In the main most of the metrics come from the statistics event that is emitted at a pre-determined time. (This will be set to every 5 seconds).

Other metrics come from specific event types using `ActiveSupport::Notifications`

### How?

The updated metrics that have been agreed between @robertomiranda, @alinagrishchuk and I are detailed below:

Karafka (Consumer metrics)

## Connection Subscriber Metrics (Global Web + Global Feeds)

| Current metric | New Metric | Done |
| - | - | - | 
| Connection latency | statistics/brokers/network.latency.avg/rtt | :white_check_mark: |
| Connection calls | statistics/brokers/connection.connects | :white_check_mark: |
| Request size | Speak to Ed/Joan | - |
| Connection response size | Speak to Ed/Joan | - |
| Connection errors | Error "connection" subscriber, tagged with "error name"| :white_check_mark: |

## Consumer Subscriber Metrics (Global Web + Global Feeds)

| Current metric | New Metric | Done |
| - | - | - |
| Consumer time lag | on_consumer_consumed / calculate time lag  | - |
| Consumer offset lag | statistics/each topic `consumer_lag_stored` |  :white_check_mark: |
| Consumer message processing latency | on_consumer_consumed/event.duration | :white_check_mark: |
| Consumer messages processed | on_consumer_consumed/messages.count (increment) | :white_check_mark: |
| Consumer messages fetched | TBC | N/A |
| Consumer message processing errors | Error "consumer" subscriber, tagged with "error name" | :white_check_mark: |
| Consumer batch processing latency | N/A  | N/A |
| Consumer batch processing errors | N/A  | N/A |
| Consumer join group | N/A | N/A |
| Consumer join group errors | Error "consumer", tagged with "error name" | :white_check_mark: |
| Consumer sync group | N/A | N/A |
| Consumer sync group errors | Error "consumer", tagged with "error name"  | :white_check_mark: |
| Consumer leave group | N/A | N/A |
| Consumer leave group errors | Error "consumer", tagged with "error name"  | :white_check_mark: |
| Consumer pause duration | N/A  | N/A |

Things to note: 

All errors are handled in the same way and increment a counter and add a label with the `error[:type]`.

### CODE REVIEW

File: lib/promenade/karafka/error_subscriber.rb
Comment (robertomiranda) [inline lib/promenade/karafka/error_subscriber.rb:16]:
```suggestion
```

File: lib/promenade/karafka/statistics_subscriber.rb
Comment (dannyd4315) [reply lib/promenade/karafka/statistics_subscriber.rb:73]: Yes, this will be removed once we have approval. All of the logging will be removed 👍

File: lib/promenade/karafka/consumer_subscriber.rb
Comment (dannyd4315) [reply lib/promenade/karafka/consumer_subscriber.rb:8]: This is microseconds

File: lib/promenade/karafka/statistics_subscriber.rb
Comment (robertomiranda) [inline lib/promenade/karafka/statistics_subscriber.rb:33]: nitpick looks like this method only needs to know about topics array:
```suggestion
        def report_topic_metrics(topics, group)
          topics.map do |topic_name, topic_values|
```

```ruby
report_topic_metrics(statistics[:topics], group)
```

File: lib/promenade/karafka/statistics_subscriber.rb
Comment (robertomiranda) [inline lib/promenade/karafka/statistics_subscriber.rb:63]: what is this label for?

File: lib/promenade/karafka/statistics_subscriber.rb
Comment (robertomiranda) [inline lib/promenade/karafka/statistics_subscriber.rb]: Similar to `report_connection_metrics` we can reduce the knowledge of this method, looks like it only needs to handle client_id and brokers values

```suggestion
        def report_connection_metrics(client_id, broker_statistics)
```

```ruby
report_connection_metrics(statistics[:client_id], statistics[:brokers])
```

File: lib/promenade/karafka/statistics_subscriber.rb
Comment (dannyd4315) [reply lib/promenade/karafka/statistics_subscriber.rb:11]: I agree option 2 is best. I have updated. 

I will also update all the `docs` within the promenade metric definitions with the specific unit of measurement.

Comment (robertomiranda) [reply lib/promenade/karafka/statistics_subscriber.rb:11]: Nice thanks Danny 🙇

File: lib/promenade/karafka/consumer_subscriber.rb
Comment (errm) [inline lib/promenade/karafka/consumer_subscriber.rb:9]:
```suggestion
        doc "Consumer message processing latency in seconds"
```

Best practice is to always use base units ... so for time that is seconds!

File: lib/promenade/karafka/consumer_subscriber.rb
Comment (errm) [inline lib/promenade/karafka/consumer_subscriber.rb:8]: Assuming that this metric is measuring the length of time it takes to process a batch, I think `duration` is a better name than `latency` - which  has a slightly different meaning to my understanding...

🤷 it could be I just don't understand what we are measuring with this metric though :)

```suggestion
      Promenade.histogram :kafka_consumer_batch_processing_duration do
```

File: lib/promenade/karafka/statistics_subscriber.rb
Comment (errm) [reply lib/promenade/karafka/statistics_subscriber.rb:11]: I think we should use seconds here ... if the old metric was incorrectly using different units we can just switch to `kafka_connection_latency_seconds` which I think is best practice anyway!

File: lib/promenade/karafka/consumer_subscriber.rb
Comment (dannyd4315) [reply lib/promenade/karafka/consumer_subscriber.rb:9]: I have updated, the previous (same) metric, was measured in miliseconds

File: lib/promenade/karafka/consumer_subscriber.rb
Comment (errm) [inline lib/promenade/karafka/consumer_subscriber.rb:20]: I can't find the place in the karafka source where this event is emmited so I can't double check.

But if the units are microseconds then this needs to be:

```suggestion
        batch_processing_duration = event.payload[:time] / 1000000.0
```

📝 note also that you need to devide by a float here or you essentially ignore most of the buckets in the histogram

File: lib/promenade/karafka/statistics_subscriber.rb
Comment (errm) [reply lib/promenade/karafka/statistics_subscriber.rb:11]: > I will also update all the docs within the promenade metric definitions with the specific unit of measurement.

Also don't worry about the `doc` too much, it is essentially ignored, so is basically just a comment, getting the name of the metric correct is much more important!

File: lib/promenade/karafka/consumer_subscriber.rb
Comment (dannyd4315) [reply lib/promenade/karafka/consumer_subscriber.rb:20]: The units are currently in miliseconds 👍 

Will update to divide by a float, thanks!

File: lib/promenade/karafka/statistics_subscriber.rb
Comment (dannyd4315) [reply lib/promenade/karafka/statistics_subscriber.rb]: Removed as agreed 👍

**Comment by dannyd4315** [comment]
@errm We have some issues with specs on Promenade around labels. We are generally trying to match the old metric names with the same units/measurements, however in quite a few cases we do not have the previous "labels" available to us when instrumenting.

I guess we would need to redefine a new metric name with the new labels?

As an example:
```
4) Promenade::Karafka statistics.karafka reports connection_metrics exposes the kafka connection calls
     Failure/Error: Promenade.metric(:kafka_connection_calls).increment(labels.merge(broker: broker_name), connection_calls)

     Prometheus::Client::LabelSetValidator::InvalidLabelSetError:
       labels must have the same signature: (expected keys: [:api, :broker, :client], got: [:broker, :client])
```

Do you think thats ok? Or a preference of keeping the metric name and adding a nil value for the labels that are not available?

**Comment by dannyd4315** [comment]
> @errm We have some issues with specs on Promenade around labels. We are generally trying to match the old metric names with the same units/measurements, however in quite a few cases we do not have the previous "labels" available to us when instrumenting.
> 
> I guess we would need to redefine a new metric name with the new labels?
> 
> As an example:
> 
> ```
> 4) Promenade::Karafka statistics.karafka reports connection_metrics exposes the kafka connection calls
>      Failure/Error: Promenade.metric(:kafka_connection_calls).increment(labels.merge(broker: broker_name), connection_calls)
> 
>      Prometheus::Client::LabelSetValidator::InvalidLabelSetError:
>        labels must have the same signature: (expected keys: [:api, :broker, :client], got: [:broker, :client])
> ```
> 
> Do you think thats ok? Or a preference of keeping the metric name and adding a nil value for the labels that are not available?

I have made an executive decision to stick with the original metric names and add and empty string for the "api" label. Makes most sense to me

**Comment by dannyd4315** [comment]
@errm @robertomiranda Please can you approve the PR and we can then deploy feeds?

File: lib/promenade/karafka/statistics_subscriber.rb
Comment (robertomiranda) [inline lib/promenade/karafka/statistics_subscriber.rb:67]: nitpick: this might be easier to forget. Could we add some comments about what this line is doing?

converting microseconds to seconds

Alternatively, you could capture this into a private method
```suggestion
            rtt = convert_microseconds_to_seconds(broker_values[:rtt][:avg])
```

**Review by robertomiranda — APPROVED** [review]
LGTM thanks @dannyd4315 🚀

**Pull request merged**
